File: typescript/packages/management/src/tools/update_source.ts
import type {
  DataSourceIngestionMode,
  UpdateDataSourceRequest,
} from '@copass/core';

import type { ToolContext, ToolHandler } from '../registrar.js';

export const updateSource: ToolHandler = async (
  ctx: ToolContext,
  input: Record<string, unknown>,
) => {
  const sourceId = String(input.data_source_id);
  const updates: UpdateDataSourceRequest = {};
  if (typeof input.name === 'string') {
    updates.name = input.name;
  }
  if (typeof input.ingestion_mode === 'string') {
    updates.ingestion_mode = input.ingestion_mode as DataSourceIngestionMode;
  }
  if (typeof input.external_account_id === 'string') {
    updates.external_account_id = input.external_account_id;
  }
  if (typeof input.poll_interval_seconds === 'number') {
    updates.poll_interval_seconds = input.poll_interval_seconds;
  }

  let mergeAdapterConfig = false;
  let adapterConfig: Record<string, unknown> | undefined;
  if (
    typeof input.adapter_config === 'object' &&
    input.adapter_config !== null
  ) {
    adapterConfig = { ...(input.adapter_config as Record<string, unknown>) };
  }
  if (typeof input.ingest_to_graph === 'boolean') {
    adapterConfig = { ...(adapterConfig ?? {}), ingest_to_graph: input.ingest_to_graph };
    mergeAdapterConfig = true;
  }
  if (adapterConfig !== undefined) {
    updates.adapter_config = adapterConfig;
  }

  const source = await ctx.client.sources.update(
    ctx.sandboxId,
    sourceId,
    updates,
    { mergeAdapterConfig },
  );
  return { source };
};

File: python/copass-management/src/copass_management/tools/update_source.py
from __future__ import annotations

from typing import Any, Dict, TYPE_CHECKING

if TYPE_CHECKING:
    from copass_management.registrar import ToolContext


async def update_source(ctx: "ToolContext", input: Dict[str, Any]) -> Any:
    source_id = str(input["data_source_id"])
    kwargs: Dict[str, Any] = {}
    if "name" in input and input["name"] is not None:
        kwargs["name"] = str(input["name"])
    if "ingestion_mode" in input and input["ingestion_mode"] is not None:
        kwargs["ingestion_mode"] = str(input["ingestion_mode"])
    if "external_account_id" in input and input["external_account_id"] is not None:
        kwargs["external_account_id"] = str(input["external_account_id"])
    if "poll_interval_seconds" in input and input["poll_interval_seconds"] is not None:
        kwargs["poll_interval_seconds"] = int(input["poll_interval_seconds"])

    merge_adapter_config = False
    adapter_config: Dict[str, Any] = {}
    if isinstance(input.get("adapter_config"), dict):
        adapter_config.update(input["adapter_config"])
    if "ingest_to_graph" in input and input["ingest_to_graph"] is not None:
        adapter_config["ingest_to_graph"] = bool(input["ingest_to_graph"])
        merge_adapter_config = True
    if adapter_config:
        kwargs["adapter_config"] = adapter_config

    source = await ctx.client.sources.update(
        ctx.sandbox_id,
        source_id,
        merge_adapter_config=merge_adapter_config,
        **kwargs,
    )
    return {"source": source}
